File: src/main/java/org/virtuoso/escape/model/data/DataLoader.java
package org.virtuoso.escape.model.data;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import org.virtuoso.escape.model.Difficulty;
import org.virtuoso.escape.model.account.Score;

import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

/**
 * Loads data from JSON files.
 *
 * @author Bose
 * @author Treasure
 * @author Andrew
 */
public class DataLoader {
    public static String ACCOUNTS_PATH = "accounts.json";
    public static String LANGUAGE_PATH = "language.json";
    public static String GAMESTATES_PATH = "gamestates.json";
    /**
     * Load all accounts from accounts.json
     *
     * @return An id-account accounts map.
     */
    public static JSONObject loadAccounts() {
        JSONObject result = new JSONObject();
        JSONObject root = parseJsonFile(Path.of("json", DataLoader.ACCOUNTS_PATH));
        if (root == null) return result;
        for (Object key : root.keySet()) {
            String id = String.valueOf(key);
            Object value = root.get(id);
            if (value instanceof JSONObject acct) {
                String username = acct.get("username").toString();
                String hashed = acct.get("hashedPassword").toString();
                Object highScore = acct.get("highScore");
				Object ttsOn = acct.getOrDefault("ttsOn", true);
                if (highScore instanceof JSONObject)
                    result.put(id, new JSONObject(Map.of(
                            "username", username,
                            "hashedPassword", hashed,
                            "highScore", highScore,
							"ttsOn", ttsOn
                    )));
            }
        }
        return result;
    }

    /**
     * Loads language mapping from language.json
     *
     * @return A mapping of id: (id: string)
     */
    public static Map<String, Map<String, String>> loadGameLanguage() {
        Map<String, Map<String, String>> result = new HashMap<String, Map<String, String>>();
        JSONObject root = parseJsonFile(Path.of("json", LANGUAGE_PATH));
        if (root == null) return result;
        for (Object key : root.keySet()) {
            String id = String.valueOf(key);
            Object inner = root.get(id);
            if (inner instanceof JSONObject innerObj) {
                Map<String, String> map = new HashMap<String, String>();
                for (Object sKey : innerObj.keySet()) {
                    String sk = String.valueOf(sKey);
                    Object sval = innerObj.get(sk);
                    if (sval != null) map.put(sk, String.valueOf(sval));
                }
                result.put(id, map);
            }
        }
        return result;
    }

    /**
     * Loads all high scores from a file.
     *
     * @return A id-score mapping.
     */
    public static Map<String, Score> loadHighScores() {
        Map<String, Score> result = new HashMap<>();
        JSONObject root = parseJsonFile(Path.of("json", ACCOUNTS_PATH));
        if (root == null) return result;

        for (Object key : root.keySet()) {
            String id = String.valueOf(key);
            Object value = root.get(id);
            if (value instanceof JSONObject acct) {
                Object high = acct.get("highScore");
                if (high instanceof JSONObject score) {
                    long seconds = (long) score.get("timeRemaining");
                    Difficulty difficulty = Difficulty.valueOf(score.get("difficulty").toString());
                    Score highScore = new Score(Duration.ofSeconds(seconds), difficulty);
                    result.put(id, highScore);
                }
            }
        }
        return result;
    }

    /**
     * Load all gamestates from gamestates.json
     *
     * @return An id-gamestate mapping.
     */
    public static JSONObject loadGameStates() {
        JSONObject result = new JSONObject();
        JSONObject root = parseJsonFile(Path.of("json", GAMESTATES_PATH));
        if (root == null) return result;
        for (Object key : root.keySet()) {
            String id = String.valueOf(key);
            Object inner = root.get(id);
            if (inner instanceof JSONObject innerObj) {
                JSONObject map = new JSONObject();
                for (Object sKey : innerObj.keySet()) {
                    String sk = String.valueOf(sKey);
                    Object sval = innerObj.get(sk);
                    if (sval != null) {
                        if (sk.equals("currentItems")) {
                            sval = innerObj.get(sk);
                            map.put(sk, sval);
						} else if (sk.equals("completedPuzzles")) {
                            sval = innerObj.get(sk);
                            map.put(sk, sval);
                        } else if (sk.equals("time")) {
                            sval = innerObj.get(sk);
                            map.put(sk, sval);
                        } else if (sk.equals("currentEntityStates")) {
                            sval = innerObj.get(sk);
                            map.put(sk, sval);
						 } else if (sk.equals("hintsUsed")) {
                            sval = innerObj.get(sk);
                            map.put(sk, sval);
                        } else map.put(sk, String.valueOf(sval));
                    }
                }
                result.put(id, map);
            }
        }
        return result;
    }

    /**
     * Parses a JSON file into a {@link JSONObject}.
     *
     * @param file The file to parse
     * @return The loaded into a {@link JSONObject}.
     */
    private static JSONObject parseJsonFile(Path file) {
        if (!Files.exists(file)) {
            System.err.println("File not found");
            return null;
        }
        JSONParser parser = new JSONParser();
        try (FileReader reader = new FileReader(file.toFile())) {
            Object obj = parser.parse(reader);
            if (obj instanceof JSONObject) return (JSONObject) obj;
            System.err.println("Not viable");
        } catch (IOException | ParseException e) {
            System.err.println("Could not parse file");
        }
        return null;
    }
}

File: src/commonMain/kotlin/org.virtuoso.escape.model/data/DataLoader.kt
@file:OptIn(ExperimentalUuidApi::class)

package org.virtuoso.escape.model.data

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.Serializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import org.virtuoso.escape.model.GameState
import org.virtuoso.escape.model.Language
import org.virtuoso.escape.model.account.Account
import org.virtuoso.escape.model.account.Score
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid


/**
 * Loads data from JSON files.
 *
 * @author Bose
 * @author Treasure
 * @author Andrew
 */
object DataLoader {
    var ACCOUNTS_PATH: String = "json/accounts.json"
    var LANGUAGE_PATH: String = "json/language.json"
    var GAMESTATES_PATH: String = "json/gamestates.json"
    var FILE_READER: ((String) -> String)? = null

    /**
     * Load all accounts from accounts.json
     *
     * @return An id-account accounts map.
     */

    /**
     * Load all accounts from accounts.json
     *
     * @return An id-account accounts map.
     */
    fun loadAccounts(): Map<Uuid, Account> {
        val root = getJsonFile(ACCOUNTS_PATH) ?: return mutableMapOf()
        return Json.decodeFromJsonElement<Map<Uuid, Account>>(root)
    }

    /**
     * Loads language mapping from language.json
     *
     * @return A mapping of id: (id: string)
     */
    fun loadGameLanguage(): Language? {
        val root = getJsonFile(LANGUAGE_PATH) ?: return null
        return Language(Json.decodeFromJsonElement<Map<String, Map<String, String>>>(root))
    }

    /**
     * Loads all high scores from a file.
     *
     * @return A id-score mapping.
     */
    fun loadHighScores(): MutableMap<Uuid, Score> {
        val accounts = loadAccounts()
        return accounts.mapValues { it.value.highScore }.toMutableMap()
    }

    /**
     * Load all gamestates from gamestates.json
     *
     * @return An id-gamestate mapping.
     */
    fun loadGameStates(): MutableMap<Uuid, GameState> {
        val root = getJsonFile(GAMESTATES_PATH) ?: return mutableMapOf()
        return Json.decodeFromJsonElement(root)
    }

    /**
     * Parses a JSON file into a [JSONObject].
     *
     * @param path The file to parse
     * @return The loaded into a [JSONObject].
     */
    private fun getJsonFile(path: String): JsonObject? {
        return FILE_READER?.invoke(path)?.let { Json.decodeFromString(it) }
    }
}

typealias SerializableDuration = @Serializable(with= DurationAsLongSerializer::class) Duration
object DurationAsLongSerializer: KSerializer<Duration>{
    override val descriptor: SerialDescriptor
        get() = SerialDescriptor("org.virtuoso.escape.Duration", Duration.serializer().descriptor)

    override fun serialize(encoder: Encoder, value: Duration) {
        encoder.encodeLong(value.inWholeSeconds)
    }

    override fun deserialize(decoder: Decoder): Duration {
        return decoder.decodeLong().seconds
    }

}
